test/externalTests/foundry.py

```diff
@@ -0,0 +1,178 @@
+#
+# (c) 2023 solidity contributors.
+# ------------------------------------------------------------------------------
+
+import os
+import re
+from pathlib import Path
+from shutil import which
+from string import Template
+from typing import Tuple
+import requests
+
+from external_test import AVAILABLE_PRESETS
+from external_test import settings_from_preset, run_cmd
+from external_test import TestConfig, TestRunner, ExternalTest
+
+# TODO: Add hardhat runner
+# TODO: Add truffle runner
+
+class FoundryRunner(TestRunner):
+    """Configure and run Foundry-based projects"""
+
+    profile_tmpl = Template("""
+[profile.${name}]
+gas_reports = [\"*\"]
+auto_detect_solc = false
+solc = \"${solc}\"
+evm_version = \"${evm_version}\"
+optimizer = ${optimizer}
+via_ir = ${via_ir}
+[profile.${name}.optimizer_details]
+yul = ${yul}
+""")
+
+    foundryup_url = "https://raw.githubusercontent.com/foundry-rs/foundry/master/foundryup/foundryup"
+    foundry_repo = "https://github.com/foundry-rs/foundry.git"
+
+    def __init__(self, config: TestConfig, setup_fn=None, compile_fn=None, test_fn=None):
+        self.config = config
+        self.setup_fn = setup_fn
+        self.compile_fn = compile_fn
+        self.test_fn = test_fn
+        self.env = os.environ.copy()
+        # Note: the test_dir will be set on setup_environment
+        self.test_dir: Path = None
+
+    def setup_environment(self, test_dir: Path, from_src: bool = False):
+        """Configure the project build environment"""
+
+        self.test_dir = test_dir
+        os.chdir(test_dir)
+
+        print("Configuring Foundry building environment...")
+        if which("forge") is None:
+            if not which("cargo"):
+                raise RuntimeError("Cargo not found.")
+            foundry_dir = test_dir / "foundry"
+            foundry_bin_dir = foundry_dir / "bin"
+            foundry_bin_dir.mkdir(parents=True, exist_ok=True)
+            self.env.update({"FOUNDRY_DIR": foundry_dir.resolve()})
+            if from_src:
+                ExternalTest.download_project(foundry_dir, self.foundry_repo, "branch", "master")
+                os.chdir(foundry_dir)
+                run_cmd(f"cargo install --path {foundry_dir.resolve()}/cli --profile local --bins --locked --force")
+            else :
+                # Create the man directory required for the foundryup script
+                foundry_man_dir = foundry_dir / "share/man/man1"
+                foundry_man_dir.mkdir(parents=True, exist_ok=True)
+                req = requests.get(self.foundryup_url, stream=True, verify=True, timeout=10)
+                foundryup_bin = test_dir / "foundryup"
+                with open(foundryup_bin, 'wb') as f:
+                    for chunk in req.iter_content(chunk_size=512):
+                        f.write(chunk)
+                run_cmd(f"chmod +x {foundryup_bin.resolve()}")
+                run_cmd(f"{foundryup_bin.resolve()}", self.env)
```

Comment:
I think we would be better off moving this part to the CI config file. We could then also cache it instead of having to reinstall foundry for each test. With Hardhat we could not easily do that because each project pulls it via its `package.json` and managing npm dependencies is messy.

Reply:
Indeed, good idea. I will change that.